File: backend/app/schemas/onboarding.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.lifecycle import ChecklistItemStatus, ChecklistStatus, ChecklistType


class OnboardingCreateRequest(BaseModel):
    employee_id: UUID

    model_config = ConfigDict(extra="forbid")


class ChecklistItemResponse(BaseModel):
    id: UUID
    checklist_id: UUID
    task_name: str
    owner_role_id: UUID
    owner_role_name: Optional[str] = None
    status: ChecklistItemStatus
    completed_by: Optional[UUID] = None
    completed_by_name: Optional[str] = None
    completed_at: Optional[datetime] = None
    sort_order: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ChecklistResponse(BaseModel):
    id: UUID
    employee_id: UUID
    employee_name: Optional[str] = None
    type: ChecklistType
    status: ChecklistStatus
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    progress_percentage: int = 0
    total_items: int = 0
    completed_items: int = 0
    items: list[ChecklistItemResponse] = []

    model_config = ConfigDict(from_attributes=True)


class ChecklistItemUpdateRequest(BaseModel):
    status: ChecklistItemStatus

    model_config = ConfigDict(extra="forbid")


class ChecklistListResponse(BaseModel):
    checklists: list[ChecklistResponse] = []
    total: int

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/onboarding.ts
export type ChecklistItemStatus = 'pending' | 'in_progress' | 'done'

export type ChecklistStatus = 'in_progress' | 'completed'

export interface ChecklistItem {
  id: string
  checklist_id: string
  task_name: string
  owner_role_id: string
  owner_role_name?: string
  status: ChecklistItemStatus
  completed_by?: string | null
  completed_by_name?: string | null
  completed_at?: string | null
  sort_order: number
  created_at: string
}

export interface OnboardingChecklist {
  id: string
  employee_id: string
  employee_name?: string
  employee_email?: string
  employee_designation?: string
  department_name?: string
  joining_date?: string
  avatar_url?: string
  type: 'onboarding'
  status: ChecklistStatus
  completed_at?: string | null
  created_at: string
  updated_at: string
  progress_percentage: number
  total_items: number
  completed_items: number
  items: ChecklistItem[]
}

export interface OnboardingCreateRequest {
  employee_id: string
  joining_date?: string
  custom_tasks?: string[]
}

export interface ChecklistItemUpdateRequest {
  status: ChecklistItemStatus
}

export interface OnboardingListResponse {
  checklists: OnboardingChecklist[]
  total: number
}

export interface AISuggestionTask {
  task_name: string
  owner_role_name: 'hr_admin' | 'it_admin' | 'facilities' | 'manager'
  category: string
  reasoning: string
}

export interface AISuggestionResponse {
  recommended_tasks: AISuggestionTask[]
  summary: string
}

export interface OnboardingSummaryMetrics {
  total_active_onboardings: number
  pending_it_tasks: number
  pending_hr_tasks: number
  avg_completion_days: number
  completion_rate: number
}
